New version of ConducWork

Olivier7017 opened the pull request:
## Summary
New version of ConducWork

* Creation of a |ConducWork| object to compute the transport property. 
* Can be created from a |PhononWork| object or from the path of a DDB and DVDB file.
* Possibility to use kerange.
* Added an example in the flow gallery.

gmatteo commented on abipy/abio/factories.py:
```diff
@@ -1413,6 +1413,109 @@ def dfpt_from_gsinput(gs_inp, ph_ngqpt=None, qpoints=None, do_ddk=True, do_dde=T
     return multi
 
 
+def conduc_from_inputs(scf_input, nscf_input, tmesh, ddb_ngqpt, eph_ngqpt_fine, sigma_erange):
```
We usually run eph calculations with boxcutmin 1.1 and mixprec 1 to decrease memory and reduce the wall-time required to apply H1. You may want to add these two arguments as optional and provide default values. Abinit default is boxcutmin 2 and mixprec 0 but here you can also opt for the fast settings.

gmatteo commented on abipy/abio/factories.py:
```diff
@@ -1413,6 +1413,109 @@ def dfpt_from_gsinput(gs_inp, ph_ngqpt=None, qpoints=None, do_ddk=True, do_dde=T
+    extension = MultiDataset.replicate_input(nscf_input, 3)
+    multi.extend(extension)
+
+    # Modify the second nscf input to get a task that interpolate the DVDB
```
I see that in your work, there's an option to activate/deactivate the intermediate step that interpolates the DVDB and writes the data to file. In our applications, we never read the DVDB from file since it becomes really huge when you have dense q-meshes. I may decide to remove the Fortran code that reads DFPT potential from file as it makes the q-point parallelism more complicated to maintain/improve.

gmatteo commented on abipy/abio/factories.py:
```diff
@@ -1413,6 +1413,109 @@ def dfpt_from_gsinput(gs_inp, ph_ngqpt=None, qpoints=None, do_ddk=True, do_dde=T
+                      eph_ngqpt_fine=eph_ngqpt_fine,
+                      eph_task=-4,
+                      tmesh=tmesh,
+                      symsigma=1,
```
symsigma 1 is the default when eph_task == -4

gmatteo commented on abipy/eph/transportfile.py:
```diff
@@ -257,8 +257,12 @@ def read_vvdos(self, component='xx', spin=1):
     def read_vvdos_tau(self, itemp, component='xx', spin=1):
         """
         Read the group velocity density of states times lifetime for different temperatures
+<<<<<<< HEAD
```
unresolved conflict

gmatteo commented on abipy/flowtk/works.py:
```diff
@@ -1967,3 +1968,156 @@ def on_all_ok(self):
+    """
+
+    @classmethod
+    def from_phwork(cls, phwork, multi, nbr_proc=None, flow=None, withKerange=False,
```
Just a minor comment on the style. We usually try to follow PEP8 for naming conventions although I don't enforce these results via scripts.
We usually use CamelCase for classes while arguments are in snake case (words separated by underscores).

gmatteo commented on abipy/flowtk/works.py:
```diff
@@ -1967,3 +1968,156 @@ def on_all_ok(self):
+        for task in new:
+            task.set_work(new)
+
+        # Manual Paralellization since autoparal doesn't work with optdriver=8 (t2)
```
> # Manual Paralellization since autoparal doesn't work with optdriver=8 (t2)

It's in my TODO list indeed.

gmatteo commented on abipy/flowtk/works.py:
```diff
@@ -1967,3 +1968,156 @@ def on_all_ok(self):
+        return new
+
+    @classmethod
+    def from_filepath(cls, DDB, DVDB, multi, nbr_proc=None, flow=None, withKerange=False,
```
I would suggest to rename the variables using:

- DDB --> ddb_path
- DBDB --> dbdb_path

just to clarify that the method expects two strings with the file path and not a DdbFile object.

gmatteo commented on abipy/flowtk/works.py:
```diff
@@ -1967,3 +1968,156 @@ def on_all_ok(self):
+            taskNumber = 4 # We have 2 more task
+
+        if skipInter: # Without separate task for DVDB interpolation
+            new.register_task(multi[taskNumber+1], deps=[Dependency(new[taskNumber-1], "WFK"),
```
If recall correctly,

'''python
deps=[Dependency(DDB, "DDB"),  Dependency(DVDB, "DVDB")]
'''

can be replaced by 


```python
deps=[DDB, "DDB"),  DVDB, "DVDB"]
'''

Olivier7017 commented:
Thanks for the reply,  I've done a new commit to follow your recommandations. In particular, 
  - Added boxcutmin and mixprec, with default value of respectively 1.1 and 1.
  - Removed the whole DVDB interpolation step to make sure my implementation works in the next months
  - Removed symsigma=1 since it is not necessary
  - Changed variable name DDB to ddb_path, DVDB to dvdb_path and withKerange to with_kerange.
  - Solved the merge conflict
  - Simplified how the Dependency are called
  
Notes : 
- I suppose boxcutmin and mixprec are only meant to be used in the conductivity task.
- I used a dict for the dependency since I couldn't make it work as a list. The following line didn't work
`deps=[(DDB, "DDB"),  (DVDB, "DVDB")]
`

gmatteo commented:
> Added boxcutmin and mixprec, with default value of respectively 1.1 and 1.
Removed the whole DVDB interpolation step to make sure my implementation works in the next months
Removed symsigma=1 since it is not necessary
Changed variable name DDB to ddb_path, DVDB to dvdb_path and withKerange to with_kerange.
Solved the merge conflict

Ok, many thanks for the modifications.
:
> I suppose boxcutmin and mixprec are only meant to be used in the conductivity task.

These variables can be used in every eph_task that needs to compute H1|Psi0> where H1 is the DFPT Hamiltonian.
so they affect eph_task 1, 4, -4 and, to some extent, the numerical values stored in the netcdf files (e.g. SIGEPH.nc)
The calculation of mobilities/conductivities done in the rta_routine is essentially a post-processing of the data stored in the SIGEPH.nc file. In a nutshell, rta_driver does not use mixprec and boxcutmin but it will use lifetimes computed with these two variables.

> I used a dict for the dependency since I couldn't make it work as a list. The following line didn't work
deps=[(DDB, "DDB"), (DVDB, "DVDB")]

My bad. deps is indeed a dictionary so 

```python
deps={new[taskNumber-1]: "WFK", phwork: ["DDB","DVDB"]}
```

is fine (and you don't need to import Dependency directly)

Thanks for contributing to the project. I'm gonna merge this PR.
